File: events/year_2023/day_07/day_07.py
"""
Advent of Code - Year 2023 - Day 7
https://adventofcode.com/2023/day/7
"""

from collections import Counter
from collections.abc import Iterator


def preprocessing(puzzle_input: str) -> list[tuple[str, int]]:
    """
    Parses the puzzle input into a list of tuples containing card strings and their corresponding
    bet amounts.
    """
    bets = [(cards, int(bet)) for cards, bet in
            (hand.split() for hand in puzzle_input.splitlines())]
    return bets


def solver(bets: list[tuple[str, int]]) -> Iterator[int]:
    """
    Yields the winnings for a variant with no joker and one with 'J' as joker.
    """
    for joker in '_J':
        yield winnings(bets, joker)


def winnings(bets: list[tuple[str, int]], joker: str):
    """
    Calculates the total winnings based on ranked poker hands with a specified joker card.
    """
    hand_sort = []
    card_strength = {
        card:
            strength for strength, card in enumerate(joker + "23456789TJQKA".replace(joker, ''))}

    for cards, bet in bets:
        jokers = cards.count(joker)
        jcards = cards.replace(joker, '')

        hand_strength = sorted(Counter(jcards).values(), reverse=True)
        hand_strength += [0 for _ in range(5 - len(hand_strength))]
        hand_strength[0] = min(5, hand_strength[0] + jokers)

        cards_strength = [card_strength[card] for card in cards]
        hand_sort.append((hand_strength, cards_strength, bet))

    return sum(n * b for n, (_, _, b) in enumerate(sorted(hand_sort), 1))

File: events/year_2023/day_07/day_07.rs
aoc::main!();

fn preprocessing(input: &str) -> Vec<(String, usize)> {
    input
        .lines()
        .map(|hand| hand.split_once(' ').unwrap())
        .map(|(cards, bet)| (cards.to_string(), bet.parse().unwrap()))
        .collect()
}

fn part_1(bets: Vec<(String, usize)>) -> usize {
    winnings(bets, '∅')
}

fn part_2(bets: Vec<(String, usize)>) -> usize {
    winnings(bets, 'J')
}

fn winnings(bets: Vec<(String, usize)>, joker: char) -> usize {
    bets.iter()
        .map(|(cards, bet)| {
            (
                cards.chars().filter(|c| *c == joker).count(),
                cards
                    .chars()
                    .filter(|c| *c != joker)
                    .counts()
                    .into_values()
                    .sorted()
                    .rev()
                    .collect_vec(),
                cards
                    .chars()
                    .map(|c| {
                        format!("{} {}", joker, "23456789TJQKA".replace(joker, ""))
                            .chars()
                            .position(|cc| cc == c)
                            .unwrap()
                    })
                    .collect_vec(),
                bet,
            )
        })
        .map(|(jokers, hand_strength, cards_strength, bet)| {
            (
                5.min(jokers + hand_strength.iter().nth(0).unwrap_or(&0)),
                hand_strength.into_iter().skip(1).collect_vec(),
                cards_strength,
                bet,
            )
        })
        .sorted()
        .enumerate()
        .map(|(n, (_, _, _, b))| (n + 1) * b)
        .sum()
}
